src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -101,7 +173,93 @@ static ucs_status_t uct_cuda_ipc_open_memhandle(const uct_cuda_ipc_rkey_t *key,
 {
     CUresult cuerr;
     ucs_status_t status;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationHandleType fabric_type = CU_MEM_HANDLE_TYPE_FABRIC;
+    CUmemoryPool *imported_mpool          = 0;
+    int key_present;
+    CUdeviceptr dptr;
+    CUmemAccessDesc access_desc;
+    CUmemGenericAllocationHandle handle;
+
+    ucs_trace("key handle type %u", key->ph.handle_type);
+
+    if (key->ph.handle_type == UCT_CUDA_IPC_KEY_HANDLE_TYPE_LEGACY) {
+        cuerr = cuIpcOpenMemHandle(mapped_addr, key->ph.handle.legacy,
+                                   CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS);
+        if (cuerr == CUDA_SUCCESS) {
+            status = UCS_OK;
+        } else {
+            ucs_debug("cuIpcOpenMemHandle() failed: %s",
+                      uct_cuda_base_cu_get_error_string(cuerr));
+            status = (cuerr == CUDA_ERROR_ALREADY_MAPPED) ?
+                     UCS_ERR_ALREADY_EXISTS : UCS_ERR_INVALID_PARAM;
+        }
+    } else if (key->ph.handle_type == UCT_CUDA_IPC_KEY_HANDLE_TYPE_VMM) {
```

Comment:
Could you please add two functions. E.g.: `uct_cuda_ipc_open_memhandle_vmm`, and `uct_cuda_ipc_open_memhandle_mempool`. And put appropriate code blocks into the methods.